JoinColony/colonyCDapp
# feat: finalize expenditure via permissions as a non owner

## Description

Implemented a saga which either finalizes the expenditure if the user is the creator, otherwise it gets the caller's arbitration permissions and tries to finish a transaction.

## Testing

You will need to operate with 2 users here, so we can test the permissions part properly.

1. Install the Staked expenditure extension
2. Give the user `fry` or `amy` Payer permissions from `leela` 
3.  create an expenditure with something like 100 as the transaction amount. Check the console for its `id`
4. fund the expenditure as leela (can also do it with the Payer user)
5. On a different browser, login as `fry` (dev wallet 3) or `amy` (dev wallet 2). NOTE: you may need to get the invite code and have them join. Just run
```
getColony(id: "0xa6Ee101fe6D2C220e373B2D1DE70A905Aa03bf41") {
    colonyMemberInviteCode
  }
```
And then go to `/invite/planex/{colonyMemberInviteCode}`
6. Enter the expenditure id and click finalize.
7. Check if the expenditure status is set to finalized by running 
```
query ListExpenditures {
  listExpenditures(filter: {nativeId: {eq: YOUR_EXPENDITURE_ID }}) {
    items {
      nativeId
      status
      finalizedAt
      ownerAddress
      balances {
        tokenAddress
        amount
      }
    }
  }
}
```
Also try to finalize a draft expenditure which is not funded as a non owner. It should stay in `DRAFT` state.
## Diffs

**Changes** 🏗

* `finalizeExpenditure` saga split into two parts, one runs for owners, the other for users with permissions

Resolves #1977
